perception/autoware_tensorrt_bevdet/include/autoware/tensorrt_bevdet/bevdet.hpp

```diff
@@ -0,0 +1,239 @@
+    buffer_num = 0;
+    init = false;
+  }
+
+  void saveFrameBuffer(
+    const void * curr_buffer, const std::string & curr_token,
+    const Eigen::Quaternion<float> & _ego2global_rot,
+    const Eigen::Translation3f & _ego2global_trans)
+  {
+    int iters = init ? 1 : n;
+    while (iters--) {
+      last = (last + 1) % n;
+      CHECK_CUDA(cudaMemcpy(
+        reinterpret_cast<char *>(adj_buffer) + last * buf_size, curr_buffer, buf_size,
+        cudaMemcpyDeviceToDevice));
+      scenes_token[last] = curr_token;
+      ego2global_rot[last] = _ego2global_rot;
+      ego2global_trans[last] = _ego2global_trans;
+      buffer_num = std::min(buffer_num + 1, n);
+    }
+    init = true;
+  }
+  int havingBuffer(int idx) { return static_cast<int>(idx < buffer_num); }
+
+  const void * getFrameBuffer(int idx)
+  {
+    idx = (-idx + last + n) % n;
+    return reinterpret_cast<char *>(adj_buffer) + idx * buf_size;
+  }
+  void getEgo2Global(
+    int idx, Eigen::Quaternion<float> & adj_ego2global_rot,
+    Eigen::Translation3f & adj_ego2global_trans)
+  {
+    idx = (-idx + last + n) % n;
+    adj_ego2global_rot = ego2global_rot[idx];
+    adj_ego2global_trans = ego2global_trans[idx];
+  }
+
+  ~adjFrame() { CHECK_CUDA(cudaFree(adj_buffer)); }
+
+private:
+  int n;
+  size_t buf_size;
+
+  int last;
+  int buffer_num;
+  bool init;
+
+  std::vector<std::string> scenes_token;
+  std::vector<Eigen::Quaternion<float>> ego2global_rot;
+  std::vector<Eigen::Translation3f> ego2global_trans;
+
+  void * adj_buffer;
+};
+
+class BEVDet
+{
+public:
+  BEVDet() {}
+  BEVDet(
+    const std::string & config_file, int n_img, std::vector<Eigen::Matrix3f> _cams_intrin,
+    std::vector<Eigen::Quaternion<float>> _cams2ego_rot,
+    std::vector<Eigen::Translation3f> _cams2ego_trans, const std::string & onnx_file,
+    const std::string & engine_file);
+
+  int DoInfer(
+    const camsData & cam_data, std::vector<Box> & out_detections, float & cost_time, int idx = -1);
+  ~BEVDet();
+
+protected:
+  void InitParams(const std::string & config_file);
+
+  void InitViewTransformer(
+    std::shared_ptr<int> & ranks_bev_ptr, std::shared_ptr<int> & ranks_depth_ptr,
+    std::shared_ptr<int> & ranks_feat_ptr, std::shared_ptr<int> & interval_starts_ptr,
+    std::shared_ptr<int> & interval_lengths_ptr);
+  void ExportEngine(const std::string & onnxFile, const std::string & trtFile);
+  int InitEngine(const std::string & engine_file);
+
+  int DeserializeTRTEngine(const std::string & engine_file, nvinfer1::ICudaEngine ** engine_ptr);
+
+  void MallocDeviceMemory();
+
+  void InitCamParams(
+    const std::vector<Eigen::Quaternion<float>> & curr_cams2ego_rot,
+    const std::vector<Eigen::Translation3f> & curr_cams2ego_trans,
+    const std::vector<Eigen::Matrix3f> & cams_intrin);
+
+  void GetAdjBEVFeature(
+    const std::string & curr_scene_token, const Eigen::Quaternion<float> & ego2global_rot,
+    const Eigen::Translation3f & ego2global_trans);
```

Comment:
(Haven't checked the code, but) Always add const whenever you can
```suggestion
  void GetAdjBEVFeature(
    const std::string & curr_scene_token, const Eigen::Quaternion<float> & ego2global_rot,
    const Eigen::Translation3f & ego2global_trans) const;
```

Reply:
Due to the `getAdjBEVFeature` function calls the non `const` member function `adj_frame_ptr ->reset()`, it may modify the state of the `adj_frame_ptr` object. Therefore, this function should not be declared as `const`.